replica/database.hh

```diff
@@ -1780,10 +1780,14 @@ public:
     static future<> flush_tables_on_all_shards(sharded<database>& sharded_db, std::string_view ks_name, std::vector<sstring> table_names);
     // flush all tables in a keyspace on all shards.
     static future<> flush_keyspace_on_all_shards(sharded<database>& sharded_db, std::string_view ks_name);
+
+    using wait_for_commitlog_reclaim = bool_class<class wait_for_commitlog_reclaim_tag>;
     // flush all tables on this shard.
     // Note: force_new_active_segment in the commitlog, so that
     // flushing all tables will allow reclaiming of all commitlog segments
-    future<> flush_all_tables();
+    //
+    // @param wait_for_commitlog__reclaim wait for the reclaimation of active commitlog segments
+    future<> flush_all_tables(wait_for_commitlog_reclaim = wait_for_commitlog_reclaim::no);
```

Comment:
I think we can just simply wait on pending deletes in all cases.
It was missed earlier, not done intentionally so there is no good reason for backward compatibility with the existing call sites.

@elcallio / @raphaelsc please stand us corrected on this matter.

Reply:
i am curious what is the merit of waiting on pending delete in all cases apart from a simpler implementation?

but i am for this change, because it's simpler. would like to create an extra commit to drop this parameter. as strictly speaking, to make "force_flush", "force_compaction" and "force_keyspace_compaction" wait for the pending delete is not in the scope of this changeset.

Reply:
It's a bug we found in the existing implementation so it actually should be fixed before this series, or with it, not after.

Reply:
Flushing all tables was introduced as part of 3ed6925673dea88a5fe157f6239b98d60ec1690a for unpinning the commitlog segments to facilitate earlier tombstone compaction.
If major comapction doesn't wait for the commitlog deletion it is also exposed to data resurrection since theoretically it could purge tombstones based on the assumption that commitlog would not resurrect data that they might shadow, BUT on a crash/restart scenario commitlog replay would happen since the commitlog segments weren't deleted - breaking the contract with compaction.

Hence waiting for deletion is required.
Maybe the function name does not convey its whole purpose - which is to flush and invalidate dirty commitlog segments.

Reply:
thank you. so it does fix bugs in "force_flush", "force_compaction" and "force_keyspace_compaction", i will create a commit with comment including your explanation above. since it is a fix,  it deserves a dedicated commit.

Reply:
thanks